cli/src/main/java/org/infinispan/cli/util/Utils.java

```diff
@@ -1,14 +1,87 @@
 package org.infinispan.cli.util;
 
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.ByteChannel;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.StandardOpenOption;
+import java.security.MessageDigest;
+
+import org.infinispan.cli.download.MavenSettings;
+import org.infinispan.commons.util.Util;
+
 public class Utils {
-   /**
-    * Returns null if the parameter is null or empty, otherwise it returns it untouched
-    */
-   public static String nullIfEmpty(String s) {
-      if (s != null && s.length() == 0) {
+   public static final int BUFFER_SIZE = 8192;
+
+   public static String sha256(Path path) {
+      return digest(path, "SHA-256");
+   }
+
+   public static String digest(Path path, String algorithm) {
+      try (ByteChannel channel = Files.newByteChannel(path, StandardOpenOption.READ)) {
+         MessageDigest digest = MessageDigest.getInstance(algorithm);
+         if (channel instanceof FileChannel) {
+            FileChannel fileChannel = (FileChannel) channel;
+            MappedByteBuffer byteBuffer = fileChannel.map(FileChannel.MapMode.READ_ONLY, 0, fileChannel.size());
+            digest.update(byteBuffer);
+         } else {
+            ByteBuffer bb = ByteBuffer.allocate(BUFFER_SIZE);
+            while (channel.read(bb) != -1) {
+               bb.flip();
+               digest.update(bb);
+               bb.flip();
+            }
+         }
+         return Util.toHexString(digest.digest());
+      } catch (NoSuchFileException e) {
+         return null;
+      } catch (Exception e) {
+         throw new RuntimeException(e);
+      }
+   }
+
+   public static Path downloadFile(URL url, Path dest, boolean verbose) throws IOException {
+      if (Files.exists(dest)) {
```

Comment:
Shouldn't we take into account the artifact's sha here?

For example, if file `example.tgz` is downloaded but corrupt. The initial artifact install will fail due to the wrong sha, however a subsequent attempt will result in `Files.exists` passing and the non-corrupt artifact will never be downloaded. Subsequently all install attempts will fail until the user manually clears the tmp cache.

Instead, if the file exists in the cache with the wrong sha, then the file should be re-downloaded and the old cached value discarded.

Reply:
I have added a new `-f/--force` option which allows you to force downloads, deleting cached artifacts. Obviously it does not work for local artifacts

Reply:
So what happens if the operator runs into the scenario I describe above? Do we always us `-f`, in which case the caching is never used?

Reply:
The alternative is to implement retries

Reply:
Also, the Maven stuff I took from WildFly modules doesn't even download the sha's from the remote repos.

Reply:
> The alternative is to implement retries

That works for me as that's the current approach taken by the operator. Ideally such a mechanism should exist in the CLI or image, so the operator can just provide the necessary dependency/patch args to the container.